sound/soc/sof/intel/hda-loader.c

```diff
@@ -371,6 +371,40 @@ int hda_dsp_cl_boot_firmware_iccmax(struct snd_sof_dev *sdev)
 	return ret;
 }
 
+static int hda_dsp_boot_imr(struct snd_sof_dev *sdev)
+{
+	struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
+	const struct sof_intel_dsp_desc *chip = hda->desc;
+	unsigned long mask;
+	u32 j;
+	int ret;
+
+	/* init for booting wait */
+	init_waitqueue_head(&sdev->boot_wait);
+
+	/* power up & unstall/run the cores to run the firmware */
+	ret = hda_dsp_enable_core(sdev, chip->init_core_mask);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: dsp core start failed %d\n", ret);
+		return -EIO;
+	}
+
+	/* set enabled cores mask and increment ref count for cores in init_core_mask */
+	sdev->enabled_cores_mask |= chip->init_core_mask;
+	mask = sdev->enabled_cores_mask;
+	for_each_set_bit(j, &mask, SOF_MAX_DSP_NUM_CORES)
+		sdev->dsp_core_ref_count[j]++;
```

Comment:
@keyonjie I was not able to understand the difference with the first boot where this code is used

````c
/* step 1: power up corex */
ret = hda_dsp_enable_core(sdev, chip->host_managed_cores_mask);
````

init_core_mask, enable_cores_mask, host_managed_cores_mask, that looks like a recipe for bugs.

Reply:
> @keyonjie I was not able to understand the difference with the first boot where this code is used
> 
> ```c
> /* step 1: power up corex */
> ret = hda_dsp_enable_core(sdev, chip->host_managed_cores_mask);
> ```
> 
> init_core_mask, enable_cores_mask, host_managed_cores_mask, that looks like a recipe for bugs.

It could be, but that is not introduced by this PR, no?